Assets/Scripts/Layers/Adapters/ChainOfResponsibility.cs

```diff
@@ -0,0 +1,134 @@
+                yield break;
+            }
+
+            // Find the proper adapter in a chain of responsibility
+            yield return AdapterChain(urlAndData);
+        }
+
+        /// <summary>
+        /// We download the file to a local cache so we can check the content from the adapters using streamreading.
+        /// This way we keep the heap memory usage low and can handle large files (like large obj's, or large WFS responses)
+        /// </summary>
+        /// <param name="urlAndData">The local file object where the path will set</param>
+        /// <returns></returns>
+        private IEnumerator DownloadDataToLocalCache(LocalFile urlAndData)
+        {
+            var url = urlAndData.SourceUrl;
+            var optionalExtention = Path.GetExtension(url);
+            var uwr = UnityWebRequest.Get(url);
+            var guidFilename = Guid.NewGuid().ToString() + optionalExtention;
+            string path = Path.Combine(Application.persistentDataPath, guidFilename);
+
+            uwr.downloadHandler = new DownloadHandlerFile(path);
+            yield return uwr.SendWebRequest();
+            if (uwr.result == UnityWebRequest.Result.Success)
+            {
+                urlAndData.LocalFilePath = path;
+            }
+            else
+            {
+                urlAndData.LocalFilePath = "";
+                Debug.LogError("Download failed: " + uwr.error);
+            }
+        }
+
+        private IEnumerator AdapterChain(LocalFile urlAndData)
+        {
+            // Get our interface references
+            dataTypeAdapterInterfaces = new IDataTypeAdapter[dataTypeAdapters.Length];
+            for (int i = 0; i < dataTypeAdapters.Length; i++)
+                dataTypeAdapterInterfaces[i] = dataTypeAdapters[i] as IDataTypeAdapter;
+
+            // Check data type per adapter using order set in inspector
+            foreach (var adapter in dataTypeAdapterInterfaces)
+            {
+                if (adapter.Supports(urlAndData))
+                {
+                    Debug.Log("<color=green>Adapter found: " + adapter.GetType().Name + "</color>");
+                    adapter.Execute(urlAndData);
+                    OnAdapterFound.Invoke();
+                    yield break;
+                }
+            }
+
+            CouldNotFindAdapter.Invoke(urlAndData.SourceUrl);
+        }
+
+        private void OnValidate() {
+            if(!Application.isPlaying || !gameObject.activeInHierarchy)
+                return;
+
+            // Make sure all scriptable objects we plug in are of the correct type
+            for (int i = 0; i < dataTypeAdapters.Length; i++)
+            {
+                if (dataTypeAdapters[i] is not IDataTypeAdapter)
+                {
+                    Debug.LogError("ScriptableObject does not have the IDataTypeAdapter interface implemented. Removing from chain.", dataTypeAdapters[i]);
+                    dataTypeAdapters[i] = null;
```

Comment:
I probably would have used a List and used the remove function to make there is no left-over nulls. But I assume you didn't do that for a reason?

Reply:
This way the serialized slot is still there so you can drag the correct SO in. (This could only happen if a dev tried to drag in a wrong SO without the adapter)

Reply:
Check for nulls in adapter loop